### PR TITLE
[GEOS-10587] Allow DataStore to auto disable on connection failure

### PR DESCRIPTION
**Pull request by taba90**
[GEOS-10587](https://osgeo-org.atlassian.net/browse/GEOS-10587)

This pr adds a property to DataStore in order to allow, when flagged, the autodisabling when connection fails.

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: doc/en/user/source/data/webadmin/stores.rst [hunk 1]
```diff
@@ -47,6 +47,7 @@ The basic information is common for all formats.
 * **Data Source Name** - the store name as listed on the view page
 * **Description** - (optional) a description that displays in the administration interface
 * **Enabled** - enables or disables access to the store, along with all datasets defined for it
+* **Auto disable on connection failure** - auto disables access to the store if any issue happens when geoserver try to acquire access to the store
```
Comment (aaime) [inline doc/en/user/source/data/webadmin/stores.rst:50]:
```suggestion
* **Auto disable on connection failure** - auto disables access to the store if any issue happens when GeoServer tries to acquire access to the store
```

File: src/main/src/main/java/org/geoserver/catalog/impl/StoreInfoImpl.java [hunk 2]
```diff
@@ -47,6 +47,8 @@ public abstract class StoreInfoImpl implements StoreInfo {
 
     protected Date dateModified;
 
+    protected boolean disableOnConnFailure;
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/catalog/impl/StoreInfoImpl.java:50]: New field, but equals and hashCode have not been updated.

File: src/main/src/main/java/org/geoserver/catalog/StoreInfo.java [hunk 3]
```diff
@@ -170,4 +170,18 @@ public interface StoreInfo extends CatalogInfo {
      */
     // <T extends Resource> Iterator<T> getResources(ProgressListener monitor) throws IOException;
 
+    /**
+     * Get the auto disable flag on connection failure.
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/catalog/StoreInfo.java:174]:
```suggestion
     * When true, the {@link ResourcePool} will automatically disable the store on connection failure. Defaults to false.
```

(I should have linked to the Resource

File: src/main/src/main/java/org/geoserver/security/config/BaseSecurityNamedServiceConfig.java [hunk 4]
```diff
@@ -94,6 +94,17 @@ public SecurityConfig clone(boolean allowEnvParametrization) {
                     && gsEnvironment != null
                     && GeoServerEnvironment.allowEnvParametrization()) {
                 target.setName((String) gsEnvironment.resolveValue(name));
+                /*try {
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/security/config/BaseSecurityNamedServiceConfig.java:97]: Unrelated change, please remove.

File: src/main/src/main/java/org/geoserver/catalog/ResourcePool.java [hunk 5]
```diff
@@ -582,6 +588,12 @@ public DataAccessFactory getDataStoreFactory(DataStoreInfo info) throws IOExcept
         }
     }
 
+    private void disableStoreInfo(WorkspaceInfo ws, String name) {
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/catalog/ResourcePool.java:591]: The documentation, GUI and REST configuration work covers all types of store. And yet, the ResoucePool/ResourcePoolTest changes seem to only cover DataStore implementations, I don't see the equivalent for the other types of store.

**Review by aaime — CHANGES_REQUESTED** [review]
@taba90 looks like the development was cut short and some bits are missing

**Comment by taba90** [comment]
@aaime thanks for the review and sorry the mess. I've pushed a branch with uncommitted changes.
Your feedback have been applied.

**Comment by aaime** [comment]
Looks good, thanks. Merging!

**Comment by jodygarnett** [comment]
I was not keen on this functionality/workflow - see comments on jira ticket.

**Comment by aaime** [comment]
@jodygarnett you did not want the functionality to be merged and all you said on the ticket was:

```
I thought they already did this? Perhaps only during initial connection attempt?

So what is the user experience you are trying for here …
```

Next time try to be clearer. That said, I don't think you understand what the flag is doing, or what is the default behavior of it (it's off by default).

**Pull request merged**